### PR TITLE
fix datetime issues

### PR DESCRIPTION
**Pull request by mathause**
fixes: #975
work around: pydata/pandas#14068

I finally found 3 issues:

(1) pretty print fails when date is out of bounds
(2) pandas Overflow error (https://github.com/pydata/pandas/issues/14068)
(3) decode_cf_datetime when first date is in bound but later dates are not

    import xarray as xr
    from datetime import datetime

    # (1) pretty print
    xr.core.formatting.format_timestamp(datetime(2800, 1, 1))

    # (2) overflow error
    ds = xr.Dataset(coords={'time' : [0, 266 * 365]})
    units = 'days since 2000-01-01 00:00:00'
    ds.time.attrs = dict(units=units)
    ds_decoded = xr.conventions.decode_cf(ds)
    ds_decoded.time

    import netCDF4 as nc
    import numpy as np
    import xarray as xr

    # (3) create netCDF file
    ncf = nc.Dataset('test_future.nc', 'w')
    ncf.createDimension('time')
    ncf.createVariable('time', np.int, dimensions=('time'))
    ncf.variables['time'].units = 'days since 1850-01-01 00:00:00'
    ncf.variables['time'].calendar = 'standard'
    ncf.variables['time'][:] = np.arange(850) * 365
    ncf.close()

    # open with xr
    ds = xr.open_dataset('test_future.nc')
    # this works
    ds
    # ds.time is a datetime64[ns] object
    # this fails
    ds.time

### CODE REVIEW

File: xarray/core/formatting.py [hunk 1]
```diff
@@ -82,10 +83,24 @@ def first_n_items(x, n_desired):
...
+
 
 def format_timestamp(t):
     """Cast given object to a Timestamp and return a nicely formatted string"""
-    datetime_str = unicode_type(pd.Timestamp(t))
+	# Timestamp is only valid for 1678 to 2262
+    try:
+        datetime_str = unicode_type(pd.Timestamp(t))
+    except OutOfBoundsDatetime:
+        datetime_str = unicode_type(t.__str__())
```
Comment (shoyer) [inline xarray/core/formatting.py]: just use `unicode_type(t)`

File: xarray/core/formatting.py [hunk 2]
```diff
@@ -82,10 +83,24 @@ def first_n_items(x, n_desired):
...
+        return []
+
+    indexer = (slice(-1, None), ) * x.ndim
+    return x[indexer]
+
 
 def format_timestamp(t):
     """Cast given object to a Timestamp and return a nicely formatted string"""
-    datetime_str = unicode_type(pd.Timestamp(t))
+	# Timestamp is only valid for 1678 to 2262
```
Comment (shoyer) [inline xarray/core/formatting.py]: too much indentation

File: xarray/conventions.py [hunk 3]
```diff
@@ -369,10 +375,17 @@ def __init__(self, array, units, calendar=None):
...
+
+        # Verify that at least the first and last date can be decoded 
+        # successfully. Otherwise, tracebacks end up swallowed by 
+        # Dataset.__repr__ when users try to view their lazily decoded array.
+        example_value = first_n_items(array, 1) or [0]
+
+        if array.size > 1:
+            # fixes (part of) https://github.com/pydata/xarray/issues/975
+            example_value_end = last_item(array)
+            example_value = np.concatenate((example_value, example_value_end))
```
Comment (shoyer) [inline xarray/conventions.py]: Rather than going through all these conditions, I would just construct this once, e.g.,
```
example_values = np.concatenate([first_n_items(array, 1), last_item(array), [0]])
```

It's not very expensive to try decoding even three values.

File: xarray/test/test_conventions.py [hunk 4]
```diff
@@ -186,6 +186,35 @@ def test_cf_datetime(self):
                             pd.Index(actual), units, calendar)
                         self.assertArrayEqual(num_dates, np.around(encoded, 1))
 
+    def test_decode_cf_datetime_overflow(self):
```
Comment (shoyer) [inline xarray/test/test_conventions.py]: You need to mark these tests as requiring netCDF4 using `@requires_netCDF4`.

File: xarray/test/test_formatting.py [hunk 5]
```diff
@@ -74,35 +84,47 @@ def test_format_items(self):
             actual = ' '.join(formatting.format_items(item))
             self.assertEqual(expected, actual)
 
-
+
```
Comment (shoyer) [inline xarray/test/test_formatting.py]: please remove these unrelated formatting changes

**Comment by mathause** [comment]
thanks for your feedback! I updated the pr

**Comment by shoyer** [comment]
This looks great to me -- thanks!

Once you add a bug-fix note to "What's New" this will be good go in.

**Comment by shoyer** [comment]
Thanks @mathause !
